overlord/state/change.go

```diff
@@ -282,32 +294,127 @@ func init() {
 	}
 }
 
+func (c *Change) isTaskWaiting(visited map[string]taskWaitComputeStatus, t *Task, deps []*Task) bool {
+	taskID := t.ID()
+	// Retrieve the compute status of the wait for the task, if not
+	// computed this defaults to 0 (taskWaitStatusNotComputed).
+	computeStatus := visited[taskID]
+	switch computeStatus {
+	case taskWaitStatusComputing:
+		// Cyclic dependency detected, return false to short-circuit.
+		return false
+	case taskWaitStatusWaiting, taskWaitStatusNotWaiting:
+		return computeStatus == taskWaitStatusWaiting
+	}
+	visited[taskID] = taskWaitStatusComputing
+
+	var isWaiting bool
+depscheck:
+	for _, wt := range deps {
```

Comment:
Great questions! All the questions are super appreciated, they're forcing me to better explain what I've done (and sometimes I confuse myself so it's always appreciated!) 

`isChangeWaiting()` should **only** ever return `true` iff it determines all tasks in Do/Undo are blocked by tasks in either of three states: `DoneStatus`, `UndoneStatus` or `WaitStatus`, if this fails, it should default to the previous logic. I updated the docs for this function to better reflect this as I feel like it may have given off the wrong intentions.

You are absolutely correct about the missing `break depscheck` after setting `isWaiting = false` in the `default:` case, this gives the expected results, it was an oversight from adding the default label, in the previous commit.

Reply:
Thanks for the explanations, this is now much clearer to me.